### ISSUE TITLE
multi_head_attention_forward produces NaN

### ISSUE THREAD

**Issue by Mrpatekful** [pytorch/pytorch#26098]
## 🐛 Bug

I am feeding a `key_padding_mask` tensor to the `multi_head_attention_forward` function, which works fine without the mask, but otherwise it produces several NaN values in the output.

## To Reproduce

```python
encoder_layer = torch.nn.TransformerEncoderLayer(d_model=512, nhead=8) transformer_encoder = torch.nn.TransformerEncoder(encoder_layer, num_layers=6)
src = torch.rand(10, 32, 512)
padding_mask = (torch.rand(32, 10) < 0.7).bool()
print(padding_mask)
out = transformer_encoder(src, src_key_padding_mask=padding_mask)
print(torch.isnan(out).any())
```

The problem is in the `multi_head_attention_forward` in line 3257, where `float('inf')` is used for filling the values before softmax, which makes the function unstable in my opinion. If i switch this value to like -1e-20 it makes the function numerically more stable than the current version. 

```python
if key_padding_mask is not None:
   attn_output_weights = attn_output_weights.view(bsz, num_heads, tgt_len, src_len)
   attn_output_weights = attn_output_weights.masked_fill(
        key_padding_mask.unsqueeze(1).unsqueeze(2),
        float('-inf'),
   )
   attn_output_weights = attn_output_weights.view(bsz * num_heads, tgt_len, src_len)
```

Also a solution for fp16 case:

```python
def neginf(dtype):
   if dtype is torch.float16:
        return -65504
    else:
        return -1e20
```

## Environment

 - PyTorch Version (e.g., 1.0): 1.2
 - OS (e.g., Linux): Ubuntu 18
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6

**Comment by fmassa** [comment]
@zhangguanheng66 can you have a look?

**Comment by zhangguanheng66** [comment]
I have seen similar issue related to `softmax` before and unfortunately that's what mathematically should be. https://github.com/pytorch/pytorch/issues/25110.

Your proposed could be a solution (BTW, you mean '-1e+20' not `-1e-20'). If you would like to submit a PR, we are happy to review and accept. We have a bunch of deterministic tests to cover the mathematical correctness. As long as the PR passes these tests, we should be fine. Thanks.

**Comment by zhangguanheng66** [comment]
you can also try something like this and see if it works.
```
attn_weights = attn_weights.float().masked_fill(
                    key_padding_mask.unsqueeze(1).unsqueeze(2),
                    float('-inf'), ).type_as(attn_weights)
```

**Comment by shawank** [comment]
@zhangguanheng66  Thank you for the reply. I tried your solution. The error still persists. Please let us know if you have any other solution in mind.

**Comment by Mrpatekful** [comment]
It turns out there was an error in my padding mask, as I marked the padded locations with 0s and the non-padded locations with 1s. This resulted in totally masked rows, which caused the softmax to output NaNs as explained in #25110.
